src/component.cpp

```diff
@@ -139,10 +139,24 @@ void Component::removeMath()
     mPimpl->mMath.clear();
 }
 
-void Component::addVariable(const VariablePtr &variable)
+bool Component::addVariable(const VariablePtr &variable)
 {
+    // Prevent adding multiple times to list.
+    auto component = shared_from_this();
+    if (component->hasVariable(variable)) {
+        return false;
+    }
+
+    // Prevent adding to multiple components.
+    bool hasOtherParent = variable->hasParent();
+    if (hasOtherParent) {
+        auto otherParent = std::dynamic_pointer_cast<Component>(variable->parent());
+        otherParent->removeVariable(variable);
+    }
```

Comment:
Why other? Why not simply:
```
if (variable->hasParent()) {
    auto parent = std::dynamic_pointer_cast<Component>(variable->parent());
    parent->removeVariable(variable);
}
```

In fact, personally, I would even have:
```
if (variable->hasParent()) {
    std::dynamic_pointer_cast<Component>(variable->parent())->removeVariable(variable);
}
```

Reply:
"other"  so as not to overload a local function name (`parent()`) with a variable name ... and while your second suggestion would work, I'd prefer to keep it in two lines so it's easier to read.

Reply:
I find it debatable whether it's easier to read. This aside, `otherParent` sounds strange to me as it gives me the impression that the variable might have two parents. I would probably go with `component` (or `componentParent`).